solr/core/src/java/org/apache/solr/search/MatchCostQuery.java

```diff
@@ -99,16 +90,16 @@ public Explanation explain(LeafReaderContext context, int doc) throws IOExceptio
       // scorer() so that we can wrap TPI
 
       @Override
-      public Scorer scorer(LeafReaderContext context) throws IOException {
+      public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOException {
         final Scorer scorer = weight.scorer(context);
         if (scorer == null) {
           return null;
         }
         final TwoPhaseIterator tpi = scorer.twoPhaseIterator();
         if (tpi == null || tpi.matchCost() == matchCost) {
-          return scorer; // needn't wrap/delegate
+          return new SolrDefaultScorerSupplier(scorer); // needn't wrap/delegate
         }
-        return new Scorer(weight) { // pass delegated weight
+        return new SolrDefaultScorerSupplier(new Scorer() { // pass delegated weight
```

Comment:
[Q] This doesn't actually use the delegated weight, as the comment suggests...is that intentional?